File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpDao.java
package com.tcssol.expensetracker.Data;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Database;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.tcssol.expensetracker.Model.Expenses;
import com.tcssol.expensetracker.Model.PersonExp;

import java.util.List;
@Dao
public interface PersonExpDao {
    @Insert
    void insertPersonExpense(PersonExp expenseEntry);
    @Query("DELETE FROM person_expenses")
    void deleteAll();
    @Query("SELECT * FROM person_expenses ORDER BY id DESC")
    LiveData<List<PersonExp>> getExpenses();
    @Query("SELECT id,name,sum(amount) as amount,type FROM person_expenses GROUP BY name,type ORDER BY name")
    LiveData<List<PersonExp>> getExpensesGrouped();
    @Query("SELECT * FROM person_expenses WHERE SUBSTR(date_created, 6, 2)=:month AND SUBSTR(date_created, 1, 4)=:year ORDER BY id DESC")
    LiveData<List<PersonExp>> getExpensesFiltered(String month,String year);
    @Query("SELECT id,name,sum(amount) as amount,type FROM person_expenses WHERE SUBSTR(date_created, 6, 2)=:month AND SUBSTR(date_created, 1, 4)=:year GROUP BY name,type ORDER BY name")
    LiveData<List<PersonExp>> getExpensesFilteredGrouped(String month,String year);
    @Query("SELECT * FROM person_expenses WHERE person_expenses.id=:id")
    LiveData<PersonExp> get(long id);
    @Update
    void update(PersonExp expense);
    @Delete
    void delete(PersonExp expense);
    @Query("SELECT * FROM person_expenses ORDER BY id ASC")
    List<PersonExp> getExpensesList();

}

File: app/src/main/java/com/tcssol/expensetracker/Data/PersonExpDao.kt
package com.tcssol.expensetracker.Data

import androidx.room.*
import com.tcssol.expensetracker.Model.PersonExp
import kotlinx.coroutines.flow.Flow

@Dao
interface PersonExpDao {

    @Insert
    suspend fun insertPersonExpense(expenseEntry: PersonExp)

    @Query("DELETE FROM person_expenses")
    suspend fun deleteAll()

    @Query("SELECT * FROM person_expenses ORDER BY id DESC")
    fun getExpenses(): Flow<List<PersonExp>>

    @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH)
    @Query("SELECT id, name, SUM(amount) AS amount, type, contact_number, mode, has_date, pending_date, note FROM person_expenses GROUP BY name, type ORDER BY name")
    fun getExpensesGrouped(): Flow<List<PersonExp>>

    @Query("SELECT * FROM person_expenses WHERE SUBSTR(date_created, 6, 2)=:month AND SUBSTR(date_created, 1, 4)=:year ORDER BY id DESC")
    fun getExpensesFiltered(month: String, year: String): Flow<List<PersonExp>>

    @SuppressWarnings(RoomWarnings.CURSOR_MISMATCH)
    @Query("SELECT id, name, SUM(amount) AS amount, type, contact_number, mode, has_date, pending_date, note FROM person_expenses WHERE SUBSTR(date_created, 6, 2)=:month AND SUBSTR(date_created, 1, 4)=:year GROUP BY name, type ORDER BY name")
    fun getExpensesFilteredGrouped(month: String, year: String): Flow<List<PersonExp>>

    @Query("SELECT * FROM person_expenses WHERE person_expenses.id=:id")
    fun get(id: Long): Flow<PersonExp?>

    @Update
    suspend fun update(expense: PersonExp)

    @Delete
    suspend fun delete(expense: PersonExp)

    @Query("SELECT * FROM person_expenses ORDER BY id ASC")
    suspend fun getExpensesList(): List<PersonExp>
}
